docs/apis/subsystems/roles.md

```diff
@@ -0,0 +1,141 @@
+
+### Context
+
+Here are the possible contexts, listed from the most general to the most specific.
+
+| CONTEXT NAME        | CONTEXT AREA       | CONTEXT LEVEL |
+|---------------------|--------------------|---------------|
+| `CONTEXT_SYSTEM`    | the whole site     | 10            |
+| `CONTEXT_USER`      | another user       | 30            |
+| `CONTEXT_COURSECAT` | a course category  | 40            |
+| `CONTEXT_COURSE`    | a course           | 50            |
+| `CONTEXT_MODULE`    | an activity module | 70            |
+| `CONTEXT_BLOCK`     | a block            | 80            |
+
+An authorized user will be able to assign an arbitrary number of roles to each user in any context.
+
+See [Roles and modules#Context](https://docs.moodle.org/dev/Roles_and_modules#Context) for more information.
+
+### Capabilities
+
+Capabilities can have the following permissions:
+
+1. `CAP_INHERIT`
+2. `CAP_ALLOW`
+3. `CAP_PREVENT`
+4. `CAP_PROHIBIT`
+
+If no permission is defined, then the capability permission is inherited from a context that is more general than the current context. If we define different permission values for the same capability in different contexts, we say that we are overriding the capability in the more specific context.
+
+### Capability conflicts
+
+Since the capabilities in each role could be different, there could be conflict in capabilities.
+
+If we set a `PROHIBIT` on a capability, it means that the capability cannot be overridden and will ALWAYS  have a permission of prevent (deny). Prohibit always wins.   For example, Jeff has a naughty student role that prohibits him from postings in any forums (for the whole site), but he's also assigned a facilitator role in "Science forum" in the course Science and Math 101. Since prohibit always wins, Jeff is unable to post in "Science forum".
+
+## Programming Interface
+
+Although the Roles system may look complicated at first glance, implementing it in Moodle code is fairly simple.
+
+- Examples of functions handling the **Roles** can be found at [`accesslib.php`](https://github.com/moodle/moodle/blob/master/lib/accesslib.php)
+
+- Moodle comes with a list of predefined roles for example Student, Teacher, Course Manager etc. A new role can be created based on
+    those predefined roles using **role archetype:**
+
+```php
+/**
+ * Returns array of all role archetypes.
+ *
+ * @return array
+ */
+function get_role_archetypes() {
+    return array(
+        'manager'        => 'manager',
+        'coursecreator'  => 'coursecreator',
+        'editingteacher' => 'editingteacher',
+        'teacher'        => 'teacher',
+        'student'        => 'student',
+        'guest'          => 'guest',
+        'user'           => 'user',
+        'frontpage'      => 'frontpage'
+    );
+}
+```
+
+- On each page you need to find the context the user is working in, using the `get_context_instance()` function.  For example, in the forum module:
```

Comment:
`get_context_instance()` is depreceated.

Better to use one of:

- `context::instance_by_id()`
- `context_[type]::instance($typeid)`